File: src/14_day.py
from copy import deepcopy


def get_polymer_rules():
    with open("14_data.txt") as fdata:
            polymers, rules_raw = fdata.read().split("\n\n")

    rules = {}
    for r in rules_raw.split("\n"):
        a, b = r.split(" -> ")
        rules[a] = b

    return polymers, rules


def generate_polymer(polymers, rules, steps=10):
    polymers_dict = {k: 0 for k in rules}

    for i in range(len(polymers) - 1):
        polymers_dict[polymers[i:i + 2]] += 1

    for _ in range(steps):
        next_polymers_dict = deepcopy({k: v for k, v in polymers_dict.items() if v > 0})
        for p, k in next_polymers_dict.items():
            polymers_dict[p[0] + rules[p]] += k
            polymers_dict[rules[p] + p[1]] += k
            polymers_dict[p] -= k

    simple_polymers_count = {}
    for p, k in polymers_dict.items():
        simple_polymers_count[p[0]] = simple_polymers_count.get(p[0], 0) + k
        simple_polymers_count[p[1]] = simple_polymers_count.get(p[1], 0) + k

    simple_polymers_count[polymers[0]] += 1
    simple_polymers_count[polymers[-1]] += 1

    result = [v // 2 for k, v in simple_polymers_count.items()]
    return max(result) - min(result)


p, r = get_polymer_rules()

print("Part 1:", generate_polymer(p, r))
print("Part 2:", generate_polymer(p, r, 40))

File: src/14_day.rs
use std::fs;
use std::collections::HashMap;

fn generate_polymer(steps: usize) -> i32 {
    let filename: &str = "./src/14_data.txt";
    let content = fs::read_to_string(filename).expect("Something went wrong reading the file");
    let mut s = content.split("\n\n");
    let polymers = s.next().unwrap();
    let rules = s.next().unwrap();

    let mut rule_dict = HashMap::<&str, &str>::new();
    let mut polymers_dict = HashMap::<&str, u32>::new();

    for r in rules.split("\n") {
        let mut s = r.split(" -> ");
        let a = s.next().unwrap();
        let b = s.next().unwrap();
        rule_dict.insert(a, b);
        polymers_dict.insert(a, 0);
    }

    for i in 0..(polymers.len() - 1) {
        let p = &polymers[i..(i + 2)];
        polymers_dict.insert(p, polymers_dict.get(p).unwrap() + 1);
    }

    for _ in 0..steps {
        let mut temp_polymers_dict = HashMap::<&str, u32>::new();
        let temp = &polymers_dict.clone();
        // for (k, v) in temp.iter() {
        //     temp_polymers_dict.insert(&k.clone(), v.clone());
        // }

        for (k, v) in polymers_dict.iter() {
            if v < &1 {
                continue
            }
            let r = rule_dict.get(k).unwrap();
            let a: String = format!("{}{}", &r.to_owned(), &k[1..2].to_owned()).to_owned();
            let b: String = format!("{}{}", &k[0..1].to_owned(), &r.to_owned()).to_owned();
            polymers_dict.insert(&*a as &str, polymers_dict.get(&*a as &str).unwrap() + v);
            polymers_dict.insert(&*b as &str, polymers_dict.get(&*b as &str).unwrap() + v);

        }
    }

    println!("{:?}", polymers_dict);


    return 0;
}

fn main() {
    generate_polymer(10);

    // println!("Part 1 total: {:?}", create_graph());  // Part 1 total: 1691
    // println!("Part 2 total: {:?}", create_graph());  // Part 2 total: 216

}